src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketTransport.cs

```diff
@@ -0,0 +1,127 @@
+
+using Microsoft.AspNetCore.Server.Kestrel.Internal.System.IO.Pipelines;
+using Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions;
+using System;
+using System.Diagnostics;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets
+{
+    internal sealed class SocketTransport : ITransport
+    {
+        private readonly SocketTransportFactory _transportFactory;
+        private readonly IEndPointInformation _endPointInformation;
+        private readonly IConnectionHandler _handler;
+        private Socket _listenSocket;
+        private Task _listenTask;
+
+        internal SocketTransport(SocketTransportFactory transportFactory, IEndPointInformation endPointInformation, IConnectionHandler handler)
+        {
+            Debug.Assert(transportFactory != null);
+            Debug.Assert(endPointInformation != null);
+            Debug.Assert(endPointInformation.Type == ListenType.IPEndPoint);
+            Debug.Assert(handler != null);
+
+            _transportFactory = transportFactory;
+            _endPointInformation = endPointInformation;
+            _handler = handler;
+
+            _listenSocket = null;
+            _listenTask = null;
+        }
+
+        public Task BindAsync()
+        {
+            if (_listenSocket != null)
+            {
+                throw new InvalidOperationException();
+            }
+
+            IPEndPoint endPoint = _endPointInformation.IPEndPoint;
+
+            var listenSocket = new Socket(endPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+
+            // Kestrel expects IPv6Any to bind to both IPv6 and IPv4
+            if (endPoint.Address == IPAddress.IPv6Any)
+            {
+                listenSocket.DualMode = true;
+            }
+
+            try
+            {
+                listenSocket.Bind(endPoint);
+            }
+            catch (SocketException e) when (e.SocketErrorCode == SocketError.AddressAlreadyInUse)
+            {
+                throw new AddressInUseException(e.Message, e);
+            }
+
+            // If requested port was "0", replace with assigned dynamic port.
+            if (_endPointInformation.IPEndPoint.Port == 0)
+            {
+                _endPointInformation.IPEndPoint = (IPEndPoint)listenSocket.LocalEndPoint;
+            }
+
+            listenSocket.Listen(512);
+
+            _listenSocket = listenSocket;
+
+            _listenTask = Task.Run(() => RunAcceptLoopAsync());
+
+            return Task.CompletedTask;
+        }
+
+        public async Task UnbindAsync()
+        {
+            if (_listenSocket != null)
+            {
+                var listenSocket = _listenSocket;
+                _listenSocket = null;
+
+                listenSocket.Dispose();
+
+                Debug.Assert(_listenTask != null);
+                await _listenTask;
```

Comment:
Yes, it can throw.

Actually, in general, I am not sure what should be done with errors from Accept.  There doesn't seem to be a way to report them today; am I missing something?  These should be rare, but they can still happen.